### PR TITLE
enhancement(codecs, kafka source): Integrate `Decoder`/`DecodingConfig` with kafka source

### PR DESCRIPTION
**Pull request by pablosichert**
Part of #7463, has been split off from #8236.

### CODE REVIEW

**Comment by binarylogic** [comment]
Nice work. Same note as the other PR...docs 😄 .

**Comment by pablosichert** [comment]
@bruceg I tagged you here for review because I'd be interested to know if there's a way to use to finalization / batch notifier without needed to allocate the temporary `events` `Vec`.

**Comment by bruceg** [comment]
> @bruceg I tagged you here for review because I'd be interested to know if there's a way to use to finalization / batch notifier without needed to allocate the temporary `events` `Vec`.

I pushed up a change ead2c36 to the branch `pablosichert/codec-sources-12-bruce` that should take you at least a step closer to what you want. Note that this change does not remove the temporary `Vec`, but you may be able to work that construct into a stream iter without the temporary.

File: src/sources/kafka.rs [hunk 1]
```diff
@@ -208,18 +197,44 @@ async fn kafka_source(
...
-                log.insert(&headers_key, Value::from(headers_map));
+
+                let msg_topic = msg.topic().to_string();
+                let msg_partition = msg.partition();
+                let msg_offset = msg.offset();
+
+                let mut events = Vec::new();
+
+                while let Ok(Some((next, _))) = decoder.decode_eof(&mut payload) {
+                    for mut event in next {
```
Comment (lukesteensen) [inline src/sources/kafka.rs:208]: If you want to avoid this `Vec`, you should be able to wrap `payload` in a `FramedRead` and get a stream you can `map`.

Comment (pablosichert) [reply src/sources/kafka.rs]: Thanks @bruceg and @lukesteensen for the suggestions! Went with `FramedRead` + `async_stream::stream` to make it work

**Review by lukesteensen — APPROVED** [review]

**Review by bruceg — APPROVED** [review]

**Pull request merged**